app/services/analytics.rb

```diff
@@ -4,7 +4,56 @@ class Analytics
   include AnalyticsEvents
   prepend Idv::AnalyticsEventsEnhancer
 
-  attr_reader :user, :request, :sp, :session, :ahoy
+  # Analytics middleware that sends the event to Ahoy
+  class AhoyMiddleware
+    attr_reader :ahoy
+
+    def initialize(ahoy: nil, request: nil)
+      @ahoy = ahoy || Ahoy::Tracker.new(request:)
+    end
+
+    def call(event)
+      event.tap do |event|
+        ahoy.track(event[:name], event[:properties])
+      end
+    end
+  end
+
+  # Analytics middleware that augments NewRelic APM trace with additional metadata.
+  class NewRelicMiddleware
+    def call(event)
+      event.tap do
+        # Tag NewRelic APM trace with a handful of useful metadata
+        # https://www.rubydoc.info/github/newrelic/rpm/NewRelic/Agent#add_custom_attributes-instance_method
+        ::NewRelic::Agent.add_custom_attributes(
+          user_id: event.dig(:properties, :user_id),
+          user_ip: event.dig(:properties, :user_ip),
+          service_provider: event.dig(:properties, :service_provider),
+          event_name: event[:name],
+          git_sha: IdentityConfig::GIT_SHA,
+        )
+      end
+    end
+  end
```

Comment:
I think ideally these would live in `app/services/analytics/` and ApplicationController would add them when it's creating its own Analytics instance